File: CogniteSdk/src/Resources/3D.cs
// Copyright 2019 Cognite AS
// SPDX-License-Identifier: Apache-2.0

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

using Oryx.Cognite;
using static Oryx.Cognite.HandlerModule;
using HttpContext = Oryx.Context<System.Net.Http.HttpResponseMessage>;

namespace CogniteSdk.Resources
{
    /// <summary>
    /// For internal use. Contains all ThreeD methods.
    /// </summary>
    public class ThreeDResource : Resource
    {
        /// <summary>
        /// Will only be instantiated by the client.
        /// </summary>
        /// <param name="authHandler">Authentication handler.</param>
        /// <param name="ctx">Context to use for the request.</param>
        internal ThreeDResource(Func<CancellationToken, Task<string>> authHandler, HttpContext ctx) : base(authHandler, ctx)
        {
        }

        /// <summary>
        /// Retrieves list of ThreeDModels matching query.
        /// </summary>
        /// <param name="query">The query filter to use.</param>
        /// <param name="token">Optional cancellation token to use.</param>
        /// <returns>List of ThreeDs matching given filters and optional cursor</returns>
        public async Task<ItemsWithCursor<ThreeDModel>> ListAsync(ThreeDModelQuery query, CancellationToken token = default)
        {
            if (query is null)
            {
                throw new ArgumentNullException(nameof(query));
            }

            var req = ThreeDs.list<ItemsWithCursor<ThreeDModel>>(query);
            return await RunAsync(req, token).ConfigureAwait(false);
        }

        /// <summary>
        /// Create ThreeDs.
        /// </summary>
        /// <param name="ThreeDs">ThreeDs to create.</param>
        /// <param name="token">Optional cancellation token.</param>
        /// <returns>Sequence of created ThreeDs.</returns>
        public async Task<IEnumerable<ThreeDModel>> CreateAsync(IEnumerable<ThreeDModelCreate> ThreeDs, CancellationToken token = default)
        {
            if (ThreeDs is null)
            {
                throw new ArgumentNullException(nameof(ThreeDs));
            }

            var req = ThreeDs.create<IEnumerable<ThreeDModel>>(ThreeDs);
            return await RunAsync(req, token).ConfigureAwait(false);
        }

        /// <summary>
        /// Retrieves information about an ThreeD given an ThreeD id.
        /// </summary>
        /// <param name="ThreeDId">The id of the ThreeD to get.</param>
        /// <param name="token">Optional cancellation token.</param>
        /// <returns>ThreeD with the given id.</returns>
        public async Task<ThreeD> GetAsync(long ThreeDId, CancellationToken token = default)
        {
            var req = ThreeDs.get<ThreeD>(ThreeDId);
            return await RunAsync(req, token).ConfigureAwait(false);
        }

        #region Delete overloads
        /// <summary>
        /// Delete multiple ThreeDs in the same project, along with all their descendants in the ThreeD hierarchy if
        /// recursive is true.
        /// </summary>
        /// <param name="query">The query of ThreeDs to delete.</param>
        /// <param name="token">Optional cancellation token.</param>
        public async Task<EmptyResponse> DeleteAsync(ThreeDDelete query, CancellationToken token = default)
        {
            if (query is null)
            {
                throw new ArgumentNullException(nameof(query));
            }

            var req = ThreeDs.delete<EmptyResponse>(query);
            return await RunAsync(req, token).ConfigureAwait(false);
        }

        /// <summary>
        /// Delete multiple ThreeDs in the same project by identity items.
        /// </summary>
        /// <param name="items">The list of ThreeDs identities to delete.</param>
        /// <param name="token">Optional cancellation token.</param>
        public async Task<EmptyResponse> DeleteAsync(IEnumerable<Identity> items, CancellationToken token = default)
        {
            if (items is null)
            {
                throw new ArgumentNullException(nameof(items));
            }

            var query = new ThreeDDelete() { Items = items };
            return await DeleteAsync(query, token).ConfigureAwait(false);
        }

        /// <summary>
        /// Delete multiple ThreeDs in the same project by internal ids.
        /// </summary>
        /// <param name="internalIds">The list of ThreeDs ids to delete.</param>
        /// <param name="token">Optional cancellation token.</param>
        public async Task<EmptyResponse> DeleteAsync(IEnumerable<long> internalIds, CancellationToken token = default)
        {
            if (internalIds is null)
            {
                throw new ArgumentNullException(nameof(internalIds));
            }

            var query = new ThreeDDelete() { Items = internalIds.Select(Identity.Create) };
            return await DeleteAsync(query, token).ConfigureAwait(false);
        }

        /// <summary>
        /// Delete multiple ThreeDs in the same project by external ids.
        /// </summary>
        /// <param name="externalIds">The list of ThreeDs ids to delete.</param>
        /// <param name="token">Optional cancellation token.</param>
        public async Task<EmptyResponse> DeleteAsync(IEnumerable<string> externalIds, CancellationToken token = default)
        {
            if (externalIds is null)
            {
                throw new ArgumentNullException(nameof(externalIds));
            }

            var query = new ThreeDDelete() { Items = externalIds.Select(Identity.Create) };
            return await DeleteAsync(query, token).ConfigureAwait(false);
        }
        #endregion

        #region Retrieve overloads
        /// <summary>
        /// Retrieves information about multiple ThreeDs in the same project. A maximum of 1000 ThreeDs IDs may be listed
        /// per request and all of them must be unique.
        /// </summary>
        /// <param name="ids">The list of ThreeDs identities to retrieve.</param>
        /// <param name="token">Optional cancellation token.</param>
        public async Task<IEnumerable<ThreeD>> RetrieveAsync(IEnumerable<Identity> ids, CancellationToken token = default)
        {
            if (ids is null)
            {
                throw new ArgumentNullException(nameof(ids));
            }

            var req = ThreeDs.retrieve<IEnumerable<ThreeD>>(ids);
            return await RunAsync(req, token).ConfigureAwait(false);
        }

        /// <summary>
        /// Retrieves information about multiple ThreeDs in the same project. A maximum of 1000 ThreeDs IDs may be listed
        /// per request and all of them must be unique.
        /// </summary>
        /// <param name="internalIds">The list of ThreeDs internal identities to retrieve.</param>
        /// <param name="token">Optional cancellation token.</param>
        public async Task<IEnumerable<ThreeD>> RetrieveAsync(IEnumerable<long> internalIds, CancellationToken token = default)
        {
            if (internalIds is null)
            {
                throw new ArgumentNullException(nameof(internalIds));
            }

            var ids = internalIds.Select(Identity.Create);
            return await RetrieveAsync(ids, token).ConfigureAwait(false);
        }

        /// <summary>
        /// Retrieves information about multiple ThreeDs in the same project. A maximum of 1000 ThreeDs IDs may be listed
        /// per request and all of them must be unique.
        /// </summary>
        /// <param name="externalIds">The list of ThreeDs internal identities to retrieve.</param>
        /// <param name="token">Optional cancellation token.</param>
        public async Task<IEnumerable<ThreeD>> RetrieveAsync(IEnumerable<string> externalIds, CancellationToken token = default)
        {
            if (externalIds is null)
            {
                throw new ArgumentNullException(nameof(externalIds));
            }

            var ids = externalIds.Select(Identity.Create);
            return await RetrieveAsync(ids, token).ConfigureAwait(false);
        }
        #endregion

        /// <summary>
        /// Retrieves a list of ThreeDs matching the given criteria. This operation does not support pagination.
        /// </summary>
        /// <param name="query">Search query.</param>
        /// <param name="token">Optional cancellation token.</param>
        /// <returns>List of ThreeDs matching given criteria.</returns>
        public async Task<IEnumerable<ThreeD>> SearchAsync (ThreeDSearch query, CancellationToken token = default )
        {
            if (query is null)
            {
                throw new ArgumentNullException(nameof(query));
            }

            var req = ThreeDs.search<IEnumerable<ThreeD>>(query);
            return await RunAsync(req, token).ConfigureAwait(false);
        }

        /// <summary>
        /// Update one or more ThreeDs. Supports partial updates, meaning that fields omitted from the requests are not
        /// changed
        /// </summary>
        /// <param name="query">The list of ThreeDs to update.</param>
        /// <param name="token">Optional cancellation token.</param>
        /// <returns>List of updated ThreeDs.</returns>
        public async Task<IEnumerable<ThreeD>> UpdateAsync (IEnumerable<ThreeDUpdateItem> query, CancellationToken token = default )
        {
            if (query is null)
            {
                throw new ArgumentNullException(nameof(query));
            }

            var req = ThreeDs.update<IEnumerable<ThreeD>>(query);
            return await RunAsync(req, token).ConfigureAwait(false);
        }
    }
}

File: Oryx.Cognite/src/3D.fs
// Copyright 2020 Cognite AS
// SPDX-License-Identifier: Apache-2.0

namespace Oryx.Cognite

open Oryx
open Oryx.Cognite

open System.Collections.Generic
open CogniteSdk

/// Various 3D HTTP handlers.

[<RequireQualifiedAccess>]
module ThreeDModels =
    [<Literal>]
    let Url = "/3d/models"

    /// <summary>
    /// Retrieves list of 3DModels matching query, and a cursor if given limit is exceeded
    /// </summary>
    /// <param name="query">The query to use.</param>
    /// <returns>List of 3DModels matching given filters and optional cursor</returns>
    let list (query: ThreeDModelQuery) (source: HttpHandler<unit>) : HttpHandler<ItemsWithCursor<ThreeDModel>> =
        source
        |> withLogMessage "3DModels:list"
        |> getWithQuery query Url

    /// <summary>
    /// Create new 3D models in the given project.
    /// </summary>
    /// <param name="items">The 3D models to create.</param>
    /// <returns>List of created 3D models.</returns>
    let create (items: ThreeDModelCreate seq) (source: HttpHandler<unit>) : HttpHandler<ThreeDModel seq> =
        source
        |> withLogMessage "3DModel:create"
        |> create items Url

    /// <summary>
    /// Delete multiple 3DModels in the same project.
    /// </summary>
    /// <param name="threeDModels">The list of 3DModels to delete.</param>
    /// <returns>Empty result.</returns>
    let delete (ids: IEnumerable<Identity>) (source: HttpHandler<unit>) : HttpHandler<EmptyResponse> =
        let items = ItemsWithoutCursor(Items = ids)

        source
        |> withLogMessage "3DModel:delete"
        |> delete items Url

    /// <summary>
    /// Retrieves information about a 3DModel in the same project.
    /// </summary>
    /// <param name="modelId">The id of the 3DModel to get.</param>
    /// <returns>3DModels with given ids.</returns>
    let retrieve (modelId: int64) (source: HttpHandler<unit>) : HttpHandler<ThreeDModel> =
        let url = Url

        source
        |> withLogMessage "3DModels:retrieve"
        |> getById modelId url

    /// Update one or more 3DModels. Supports partial updates, meaning that
    /// fields omitted from the requests are not changed. Returns list of updated 3DModels.
    let update (query: IEnumerable<UpdateItem<ThreeDModelUpdate>>) (source: HttpHandler<unit>) : HttpHandler<ThreeDModel seq> =
        source
        |> withLogMessage "3DModels:update"
        |> update query Url

[<RequireQualifiedAccess>]
module ThreeDRevisions =
    let Url = "/3d/models"

    /// <summary>
    /// Retrieves list of 3DRevisions matching query, and a cursor if given limit is exceeded
    /// </summary>
    /// <param name="modelId">The model to get revisions from.</param>
    /// <param name="query">The query to use.</param>
    /// <returns>List of 3DRevisions matching given filters and optional cursor</returns>
    let list (modelId: int64) (query: ThreeDRevisionQuery) (source: HttpHandler<unit>) : HttpHandler<ItemsWithCursor<ThreeDRevision>> =
        let url = Url +/ sprintf "%d" modelId +/ "revisions"

        source
        |> withLogMessage "3DRevisions:list"
        |> getWithQuery query url

    /// <summary>
    /// Create new 3D Revisions in the given project.
    /// </summary>
    /// <param name="modelId">The model to get revisions from.</param>
    /// <param name="items">The 3D Revisions to create.</param>
    /// <returns>List of created 3D Revisions.</returns>
    let create (modelId: int64) (items: ThreeDRevisionCreate seq) (source: HttpHandler<unit>) : HttpHandler<ThreeDRevision seq> =
        let url = Url +/ sprintf "%d" modelId +/ "revisions"

        source
        |> withLogMessage "3DRevision:create"
        |> create items url

    /// <summary>
    /// Delete multiple 3DRevisions in the same project.
    /// </summary>
    /// <param name="modelId">The model to get revisions from.</param>
    /// <param name="threeDRevisions">The list of 3DRevisions to delete.</param>
    /// <returns>Empty result.</returns>
    let delete (modelId: int64) (ids: IEnumerable<Identity>) (source: HttpHandler<unit>) : HttpHandler<EmptyResponse> =
        let url = Url +/ sprintf "%d" modelId +/ "revisions"
        let items = ItemsWithoutCursor(Items = ids)

        source
        |> withLogMessage "3DRevision:delete"
        |> delete items url

    /// <summary>
    /// Retrieves information about multiple 3DRevisions in the same project. A maximum of 1000 3DRevision IDs may be listed per
    /// request and all of them must be unique.
    /// </summary>
    /// <param name="modelId">The model to get revisions from.</param>
    /// <param name="ids">The ids of the 3DRevisions to get.</param>
    /// <returns>3DRevisions with given ids.</returns>
    let retrieve (modelId: int64) (revisionId: int64) (source: HttpHandler<unit>) : HttpHandler<ThreeDRevision> =
        let url = Url +/ sprintf "%d" modelId +/ "revisions"

        source
        |> withLogMessage "3DRevisions:retrieve"
        |> getById revisionId url

    /// <summary>
    /// Update one or more 3DRevisions. Supports partial updates, meaning that
    /// fields omitted from the requests are not changed. Returns list of updated 3DRevisions.
    /// </summary>
    /// <param name="modelId">The model to get revisions from.</param>
    /// <param name="query">The update query.</param>
    /// <returns>Corresponing revisions after applying the updates.</returns>
    let update
        (modelId: int64)
        (query: IEnumerable<UpdateItem<ThreeDRevisionUpdate>>)
        (source: HttpHandler<unit>) 
        : HttpHandler<ThreeDRevision seq> =
        let url = Url +/ sprintf "%d" modelId +/ "revisions"

        source
        |> withLogMessage "3DRevisions:update"
        |> update query url

    /// <summary>
    /// Update one or more 3DRevisions. Supports partial updates, meaning that
    /// fields omitted from the requests are not changed. Returns list of updated 3DRevisions.
    /// </summary>
    /// <param name="modelId">The model to get revisions from.</param>
    /// <param name="revisionId">The revision to get logs from.</param>
    /// <param name="query">The update query.</param>
    /// <returns>Corresponing revisions after applying the updates.</returns>
    let updateThumbnail (modelId: int64) (revisionId: int64) (fileId: int64) (source: HttpHandler<unit>) : HttpHandler<EmptyResponse> =
        let url =
            Url
            +/ sprintf "%d" modelId
            +/ "revisions"
            +/ sprintf "%d" revisionId
            +/ "thumbnail"

        let query = ThreeDUpdateThumbnailQuery(FileId = fileId)

        source
        |> withLogMessage "3DRevisions:update"
        |> postWithQuery () query url jsonOptions

    /// <summary>
    /// Retrieves list of 3DRevisionLogs matching severity.
    /// </summary>
    /// <param name="modelId">The model to get revisions from.</param>
    /// <param name="revisionId">The revision to get logs from.</param>
    /// <param name="severity">Minimum severity to retrieve (3 = INFO, 5 = WARN, 7 = ERROR).</param>
    /// <returns>List of 3DRevisions matching given filters and optional cursor</returns>
    let listLogs
        (modelId: int64)
        (revisionId: int64)
        (query: ThreeDRevisionLogQuery)
        (source: HttpHandler<unit>) 
        : HttpHandler<ItemsWithCursor<ThreeDRevisionLog>> =
        let url =
            Url
            +/ sprintf "%d" modelId
            +/ "revisions"
            +/ sprintf "%d" revisionId
            +/ "logs"

        source
        |> withLogMessage "3DRevisionLogs:list"
        |> getWithQuery query url

module ThreeDNodes =
    let Url = "/3d/models"

    /// <summary>
    /// Retrieves list of 3DNode matching query, and a cursor if given limit is exceeded
    /// </summary>
    /// <param name="modelId">The model to get nodes from.</param>
    /// <param name="query">The query to use.</param>
    /// <returns>List of 3DNode matching given query and optional cursor</returns>
    let list
        (modelId: int64)
        (revisionId: int64)
        (query: ThreeDNodeQuery)
        (source: HttpHandler<unit>) 
        : HttpHandler<ItemsWithCursor<ThreeDNode>> =
        let url =
            Url
            +/ sprintf "%d" modelId
            +/ "revisions"
            +/ sprintf "%d" revisionId
            +/ "nodes"

        source
        |> withLogMessage "3DNode:list"
        |> getWithQuery query url

module ThreeDAssetMappings =
    let Url = "3d/models"

    /// <summary>
    /// Retrieves list of 3DAssetMapping matching query, and a cursor if given limit is exceeded
    /// </summary>
    /// <param name="modelId">The model to get asset mappings from.</param>
    /// <param name="revisionId">The revision to get asset mappings from.</param>
    /// <param name="query">The query to use.</param>
    /// <returns>List of 3DAssetMapping matching given query and optional cursor</returns>
    let list
        (modelId: int64)
        (revisionId: int64)
        (query: ThreeDAssetMappingQuery)
        (source: HttpHandler<unit>) 
        : HttpHandler<ItemsWithCursor<ThreeDAssetMapping>> =
        let url =
            Url
            +/ sprintf "%d" modelId
            +/ "revisions"
            +/ sprintf "%d" revisionId
            +/ "mappings"

        source
        |> withLogMessage "3DAssetMapping:list"
        |> list query url

    /// <summary>
    /// Create new 3D AssetMapping in the given project.
    /// </summary>
    /// <param name="modelId">The model to get asset mappings from.</param>
    /// <param name="revisionId">The revision to get asset mappings from.</param>
    /// <param name="items">The 3D asset mappings to create.</param>
    /// <returns>List of created 3D AssetMappings.</returns>
    let create
        (modelId: int64)
        (revisionId: int64)
        (items: ThreeDAssetMappingCreate seq)
        (source: HttpHandler<unit>) 
        : HttpHandler<ThreeDAssetMapping seq> =
        let url =
            Url
            +/ sprintf "%d" modelId
            +/ "revisions"
            +/ sprintf "%d" revisionId
            +/ "mappings"

        source
        |> withLogMessage "3DAssetMapping:create"
        |> create items url

    /// Delete 3D AssetMappings in the given project.
    /// </summary>
    /// <param name="modelId">The model to delete asset mappings from.</param>
    /// <param name="revisionId">The revision to delete asset mappings from.</param>
    /// <param name="assetMappings">AssetMappings to delete</param>
    /// <returns>Empty result.</returns>
    let delete
        (modelId: int64)
        (revisionId: int64)
        (assetMappings: IEnumerable<Identity>)
        (source: HttpHandler<unit>) 
        : HttpHandler<EmptyResponse> =
        let url =
            Url
            +/ sprintf "%d" modelId
            +/ "revisions"
            +/ sprintf "%d" revisionId
            +/ "mappings"

        let items = ItemsWithoutCursor(Items = assetMappings)

        source
        |> withLogMessage "3DAssetMapping:delete"
        |> delete items url
